### DISCUSSION TITLE
`update_code` in sub-directory of `release_path`

### DISCUSSION THREAD

**Discussion by stephan-strate** [deployphp/deployer#3052] [General]
I would like to be able to checkout the repositories code to a sub-directory of the `release_path`.

My goal is to create a zero-downtime recipe for [Grav](https://getgrav.org/). The flat-file cms Grav is a little bit different to most other frameworks. The repository will only contain the sub-directory `user/` with custom plugins, themes, configuration and so on (see [Folder Structure](https://learn.getgrav.org/17/basics/folder-structure)). My idea is to use `composer create-project` (see [Installation](https://learn.getgrav.org/17/basics/installation#option-2-install-with-com)) in the `release_path` first and then clone the content into the sub-directory `user/`.

**Proposal**:  
Create `update_code_path` variable, which defaults to `release_path`, but also makes it configurable for special use-cases.

What do you think?

**Comment by Schrank** [comment]
I tried that with shopware to only have it in a subdir in git and it didn’t work well. You need to change everything™️. All the build, composer install, etc. It sucked so hard, I abandoned the idea.
And now I think it doesn’t makes sense as a generic solution. It adds a lot of complexity for a small edge case very few users have.

**Comment by stephan-strate** [comment]
Hmm, I don't see the problem here. Having it configurable doesn't hurt right? Because in this case the composer and build stuff needs to stick to the `release_path`. Just the content needs to be in a sub-directory.

How would you solve the update/deployment process of such frameworks? I'm not really into shopware, so I don't know what the requirements are there.

**Comment by antonmedv** [comment]
Right now you can check out subdirectories easily https://deployer.org/docs/7.x/recipe/deploy/update_code#sub_directory

**Reply by AWqxKAWERbXo** [reply]
I don't understand how this helps. :/ Yes, I'm new to Deployer 7 and I work with a ready-to-use Deployer template for a CMS. But the solution for Deployer 6 to just set a parameter for git submodules was quite easy. Where can I add sth like 

```
run("$git submodule update --init --recursive");
```

**Comment by stephan-strate** [comment]
This works if you want to checkout a sub-directory of the git repository. I want to checkout the git repository inside a sub-directory in release_path.

Or we could make use of the `--prefix` option of this git archive command.

**Comment by Sogl** [comment]
@stephan-strate Did you manage to make a recipe for Grav?

**Reply by stephan-strate** [reply]
Yes, heres the related gist: https://gist.github.com/stephan-strate/2f410549015e99a41293b5e2761d388b.

**Reply by antonmedv** [reply]
What about adding it to Deployer recipes as well?

**Reply by stephan-strate** [reply]
Currently it's tailored to my needs and quite new. My plan was it to refine it a bit more and add it when it's battle-proofed.

**Reply by Sogl** [reply]
Thx! You're on 6 or 7 deployer version? Are you launching from the root of the project or from the user folder? I need to somehow prepare the current project for the folder structure for the Deployer.

**Reply by stephan-strate** [reply]
I am on the latest rc of deployer 7.
And I am starting it from the root of the project, where user folder is located. The reason behind this is to make it possible to override files like robots.txt or .htaccess

**Reply by Sogl** [reply]
@stephan-strate I tried to use your script and got an error:
```
task grav:install

[server.host] run cd ~/public_html/site.com/releases/1 && (/usr/local/bin/ea-php74 bin/grav install)
[server.host] In UniformResourceIterator.php line 109:
[server.host]                                                   
[server.host]   Failed to open dir: plugins:// does not exist.  
[server.host]                                                   
[server.host] install [-s|--symlink] [-p|--plugin PLUGIN] [-t|--theme THEME] [-h|--help] [-q|--quiet] [-v|vv|vvv|--verbose] [-V|--version] [--ansi] [--no-ansi] [-n|--no-interaction] [--env [ENV]] [--lang [LANG]] [--] <command> [<destination>]
[server.host]  error  in deploy.php on line 55:
[server.host] exit code 1 (General error)
task deploy:failed
```

Resut root folder:
![image](https://user-images.githubusercontent.com/7556290/181003072-890adb81-4f79-4227-8cf3-877d27ef68c0.png)

`release` symlink folder. it turned out to a mix from the main Grav repo and mine with the User folder:
![image](https://user-images.githubusercontent.com/7556290/181004762-528e6a0b-0bc7-4241-87f6-254748177867.png)

I also do not understand the concept of keeping these folders shared, because the `/user` folder is constantly changing:
![image](https://user-images.githubusercontent.com/7556290/181006683-624f488f-b0bb-494b-b5e5-a6088321d223.png)

I store in git some parts of `/user` folder. My `.gitignore`:
```
# Grav Specific
config/security.yaml
env/*/config/security.yaml
data/*
!data/.*
plugins/*
!plugins/.*
themes/*
!themes/.*

# Inherited theme
!themes/mytheme4
!themes/learn4

# Lost and inherited plugins
!plugins/snappygrav
!plugins/my-snappygrav

# JS
node_modules/

# Composer
.composer
vendor/

# OS Generated
.DS_Store*
thumbs.db
Icon?
Thumbs.db
*.swp
```

My thoughts:
1. Git clone GRAV repo only on first run. The GRAV version changes not so often and conflicts are possible when a new one is released. The operation also takes time.
2. Clone `/user` folder from git repo on every run (can every release contain only `user` folder?)
3. Install missing `.dependencies` like plugins (takes time).
4. `composer install` in all plugins, themes.
5. `bin/grav clear-cache` etc

Any ideas?

**Reply by stephan-strate** [reply]
With this script, you will need to wrap your content into a `user` folder to make it work.
As I said, this script is tailored to my needs.

All the folder I marked as shared are content folders. I don't keep the content of it in git. You could use another git repository for content together with the sync plugin. This script is only for templates, plugins, ...

1. You can do this, sure. But you won't have a full zero-downtime experience with it. With my script you can fully roll back to an older release with no problem.
2. It can, you will need to symlink the release to user then. I don't know how good this works. Also it will probably require a manual setup.
3. You don't want to execute `composer install`, because Grav plugins come with composer dependencies already

I spent a lot of time to create this script. You are free to customize or change it. I am sure everyone has other needs and wishes.

**Reply by Sogl** [reply]
Two days of debugging and I made a recipe for my needs: https://gist.github.com/Sogl/6109dd405ed1420694632c63fc5108b1

Of course, there is still a lot to optimize... but it works 😄

**Reply by stephan-strate** [reply]
Thats awesome. If we gonna submit it as an official recipe. What would we do? I encountered that as a problem when I first started to think about this. Every user has some different requirements and needs. I feel like one recipe can not rule it all. Any thoughts @Sogl @antonmedv?

**Reply by Sogl** [reply]
Two situations: 
1. User put whole folder in git
2. User put `/user` folder in git.

For second we need `update_code` to subdir (subj). Also I need `--recurse-submodules` for `git clone` that was removed form v.7 (or support recursive for `git archive`). I also want to have command to `git pull` support for current release, I don't create releases for every commit. Need @antonmedv support.

I also lack some kind of Grav version check and plugins to install the same ones as on the local copy (we are now installing the latest versions).

**Reply by stephan-strate** [reply]
Situation 2. can not be implemented with deployer right now.
Specific grav versions can be installed using the `.dependencies` file and using `bin/grav install`, which is already covered in my recipe. 

All other things are kind of not wanted by deployer, I guess.

**Reply by Sogl** [reply]
In my recipe user folder in git works fine, see my code.

Ok, you can install specific version. But how to get them all in console from local (dev) before install? Like this:
```
Grav 1.7.34
Admin 1.10.3
Breadcrumbs 1.5
```

So you can then put it into `bin/gpm install <currently installed list>`

Also you can't generate `.dependencies` from console, all by hands. Maybe it support versions, I don't know.

**Reply by antonmedv** [reply]
Can’t tell what is better for grav.  I need to try to install and deploy it on my own to understand.

**Reply by stephan-strate** [reply]
Yeah, I know it works. But its fully custom and not using deployer recipes at all.

I do not know what you mean with generating from console. You can use `bin/grav install` locally as well. You could even use deployer to deploy it locally.

It gets a little off topic at this point. You are listing things that are either not supported by grav or deployer. Thats unfortunate, I agree, but will not bring us further with the recipe.

**Reply by Sogl** [reply]
Just wanted to share...

Today I created two tasks to install Grav CMS from zip-file (.git contains `/user` folder in my case). It has some hacks for cloning repo into non-empty folder with already installed plugins. In this case I don't use Deployer bare repo at all. 

```php
desc('Download Grav + Admin zip');
task('grav:download_zip', function () {
    // get by url
    run('wget -O {{release_path}}/grav_latest.zip {{grav_zip}}');
    // unzip it in same directory
    run('unzip {{release_path}}/grav_latest.zip -d {{release_path}}');
    // move all grav-admin files (+ dotfiles dotfolders) 
    run('mv {{release_path}}/grav-admin/* {{release_path}}/grav-admin/.[!.]*  {{release_path}}');
    // remove zip and extracted folder
    run('rm -r {{release_path}}/grav_latest.zip {{release_path}}/grav-admin');
    // delete all in user except plugins
    run('find {{release_path}}/user -mindepth 1 -maxdepth 1 -name plugins -prune -o -exec rm -rf {} \;');
});

desc ('Code update');
task('my:git', function () {
    cd('{{release_path}}/user');
    $git = get('bin/git');
    $repository = get('repository');
    run("$git init");
    run("$git remote add origin $repository");
    run("$git fetch");
    run("$git checkout -t origin/master -f");
    // install submodules (I have one with learn4 theme)
    run("$git submodule update --init");
});
```

The deployment has become a little faster to work.
And my idea about zip-installation improvement: https://github.com/getgrav/grav/issues/3644
